email/reporter.go

```diff
@@ -0,0 +1,231 @@
+		staticWaitGroup sync.WaitGroup
+	}
+)
+
+// NewReporter creates a new reporter.
+func NewReporter(database *database.AbuseScannerDB, creds NCMECCredentials, logger *logrus.Logger) *Reporter {
+	return &Reporter{
+		staticClient:   NewNCMECClient(creds),
+		staticDatabase: database,
+		staticLogger:   logger.WithField("module", "Reporter"),
+		staticStopChan: make(chan struct{}),
+	}
+}
+
+// Start initializes the reporter process.
+func (r *Reporter) Start() error {
+	r.staticWaitGroup.Add(1)
+	go func() {
+		r.threadedReportMessages()
+		r.staticWaitGroup.Done()
+	}()
+	return nil
+}
+
+// Stop waits for the finalizer's waitgroup and times out after one minute.
+func (r *Reporter) Stop() error {
+	close(r.staticStopChan)
+
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		r.staticWaitGroup.Wait()
+	}()
+	select {
+	case <-c:
+		return nil
+	case <-time.After(maxShutdownTimeout):
+		return errors.New("unclean reporter shutdown")
+	}
+}
+
+// threadedReportMessages will periodically fetch email messages that have not
+// been tagged as csam and have not been reported to NCMEC yet.
+func (r *Reporter) threadedReportMessages() {
+	// convenience variables
+	logger := r.staticLogger
+
+	// create a new ticker
+	ticker := time.NewTicker(reportingFrequency)
+
+	// start the loop
+	for {
+		logger.Debugln("threadedReportMessages loop iteration triggered")
+		r.reportMessages()
+
+		select {
+		case <-r.staticStopChan:
+			logger.Debugln("Reporter stop channel closed")
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// reportMessages fetches all unreported messages from the database and reports
+// them to NCMEC. Reporting only applies to emails which have been tagged with
+// CSAM, we report those messages and the content we find inside them to NCMEC.
+func (r *Reporter) reportMessages() {
+	// convenience variables
+	abuseDB := r.staticDatabase
+	logger := r.staticLogger
+
+	// fetch all unreported emails
+	toReport, err := abuseDB.FindUnreported()
+	if err != nil {
+		logger.Errorf("Failed fetching unreported emails, error %v", err)
+		return
+	}
+
+	// log unreported message count
+	numUnreported := len(toReport)
+	if numUnreported == 0 {
+		logger.Debugf("Found %v unreported messages", numUnreported)
+		return
+	}
+
+	logger.Infof("Found %v unreported messages", numUnreported)
+
+	// loop all emails and report them
+	for _, email := range toReport {
+		err := r.reportEmail(email)
+		if err != nil {
+			logger.Errorf("Failed to report email %v, error %v", email.UID, err)
+		}
+	}
+}
+
+// reportEmail will report the given email to NCMEC, it does so by interacting
+// with the NCMEC API.
+func (r *Reporter) reportEmail(email database.AbuseEmail) error {
+	// convenience variables
+	logger := r.staticLogger
+
+	// convert the abuse email into a NCMEC report
+	report, err := emailToReport(email)
+	if err != nil {
+		return err
+	}
+
+	// open the report
+	reportedAt := time.Now().UTC()
+	resp, err := r.staticClient.openReport(report)
```

Comment:
What happens when the report isn't closed? Or you try to open the same report twice?

Reply:
Well, nothing really happens, we create noise on their end. I was thinking about this when writing it as well, so I updated the flow a bit to improve it. 

I made it so we:
- try and finish an open report if we have its id
- try and finish a report if we have its id, even if something went wrong after opening it
- update the email with potential reporting errors on both the open and finish stage of the report

No need to overthink this any more than that for the time being I think. But up for review obv.